### PR TITLE
Hook-up generator and new unitary DX equipment module 

### PR DESCRIPTION
**Pull request by lymereJ**
This PR make necessary changes to get a bare-bone connection from the new unitary DX module and the generator. A new test was added to show that the generator is able to generate sets of curves that meet a targeted efficiency. Note that this is a set of functional preliminary changes, refinements will be required.

### CODE REVIEW

File: copper/data/equipment_references.json [hunk 1]
```diff
@@ -37,20 +37,20 @@
...
         }
     },
-    "UnitaryDirectExpansion": {
+    "unitarydirectexpansion": {
         "ahri_340/360": {
             "water": {
-                "aed": [80,67],
-                "aew": [67,57],
+                "aed": 80,
+                "aew": 67,
```
Comment (lymereJ) [inline copper/data/equipment_references.json:44]: Correction; These were supposed to be single values.

File: copper/data/equipment_references.json [hunk 2]
```diff
@@ -37,20 +37,20 @@
             }
         }
     },
-    "UnitaryDirectExpansion": {
+    "unitarydirectexpansion": {
```
Comment (lymereJ) [inline copper/data/equipment_references.json:40]: Change the case for consitency.

File: copper/generator.py [hunk 3]
```diff
@@ -220,6 +222,11 @@ def run_ga(self, curves, verbose=False):
                             f"Target not met after {self.max_restart} restart; No solution was found."
                         )
                         return
+                else:
+                    logging.critical(
+                        f"Target not met after {self.max_gen} generations; No solution was found."
+                    )
+                    return
```
Comment (lymereJ) [inline copper/generator.py:229]: Add error when the number of generations has been exceeded.

File: copper/generator.py [hunk 4]
```diff
@@ -259,6 +266,26 @@ def is_target_met(self):
...
+            #                    eff_type="full", unit=self.equipment.full_eff_unit
+            #                )
+            #                cap_rating = 0
+            #                if "cap-f-t" in self.vars:
+            #                    for c in self.equipment.set_of_curves:
+            #                        # set_of_curves
+            #                        if "cap" in c.out_var:
+            #                            cap_rating += abs(1 - c.get_out_reference(self.equipment))
+            else:
+                return False
```
Comment (lymereJ) [inline copper/generator.py:288]: Preliminary target check.

File: copper/unitarydirectexpansion.py [hunk 5]
```diff
@@ -122,9 +126,10 @@ def __init__(
                     "x2_min": 10,
                     "x2_max": 40,
                     "x2_norm": ect,
+                    "nbval": nb_val,
                 },
-                "eir-f-f": {"x1_min": 0, "x1_max": 1, "x1_norm": 1, "nbval": nb_val},
-                "cap-f-f": {"x1_min": 0, "x1_max": 1, "x1_norm": 1, "nbval": nb_val},
+                "eir-f-ff": {"x1_min": 0, "x1_max": 1, "x1_norm": 1, "nbval": nb_val},
+                "cap-f-ff": {"x1_min": 0, "x1_max": 1, "x1_norm": 1, "nbval": nb_val},
```
Comment (lymereJ) [inline copper/unitarydirectexpansion.py:132]: Fix typo.

File: copper/unitarydirectexpansion.py [hunk 6]
```diff
@@ -122,9 +126,10 @@ def __init__(
                     "x2_min": 10,
                     "x2_max": 40,
                     "x2_norm": ect,
+                    "nbval": nb_val,
```
Comment (lymereJ) [inline copper/unitarydirectexpansion.py:129]: Missing field.

File: copper/unitarydirectexpansion.py [hunk 7]
```diff
@@ -286,20 +292,16 @@ def get_rated_temperatures(self, alt=False):
...
-        aew = [
-            Equipment.convert_to_deg_c(t, dx_data["ae_unit"]) for t in dx_data["aew"]
-        ]
-        # Outdoor Water/Air entering
+        dx_data = equipment_references[self.type.lower()][std][self.condenser_type]
+        # Air entering indoor dry-bulb
+        aed = Equipment.convert_to_deg_c(dx_data["aed"], dx_data["ae_unit"])
+        # Air entering indoor wet-bulb
+        aew = Equipment.convert_to_deg_c(dx_data["aew"], dx_data["ae_unit"])
+        # Outdoor water/air entering
```
Comment (lymereJ) [inline copper/unitarydirectexpansion.py:300]: Change code to apply conversion to single value.

File: tests/test_unitarydirectexpansion.py [hunk 8]
```diff
@@ -31,6 +31,33 @@ def test_calc_eff_ect(self):
...
+        set_of_curves = dx_unit.generate_set_of_curves(
+            base_curves=[lib.get_set_of_curves_by_name("D208122216")],
+            tol=0.01,
+            verbose=True,
+            vars=["eir-f-t"],
+            max_gen=300,
+            max_restart=3,
+        )
+        assert len(set_of_curves) == 5
+
```
Comment (lymereJ) [inline tests/test_unitarydirectexpansion.py:60]: New test that showcases the generation process.

File: copper/data/equipment_references.json (hunk 1 above)
Comment (wanhanlong1130) [reply copper/data/equipment_references.json:44]: but, these values are from the 210/240 I think, for two points; though they are not used in the code. We just ignore the  67,57 pair?

File: copper/data/equipment_references.json (hunk 2 above)
Comment (wanhanlong1130) [reply copper/data/equipment_references.json:40]: I will check for other places.

File: copper/data/equipment_references.json (hunk 1 above)
Comment (lymereJ) [reply copper/data/equipment_references.json:44]: Do you mean from AHRI 340/360? 67/57 is used for low temperature operating cooling, not for standard operation.

image

File: copper/data/equipment_references.json (hunk 2 above)
Comment (lymereJ) [reply copper/data/equipment_references.json:40]: I recall that we had issues with that in other places. If it works, that's good, if not, that's okay.

**Pull request merged**